src/Symfony/Component/Serializer/Serializer.php

```diff
@@ -51,59 +50,70 @@ class Serializer implements SerializerInterface, NormalizerInterface, Denormaliz
      */
     public const EMPTY_ARRAY_AS_OBJECT = 'empty_array_as_object';
 
-    private const SCALAR_TYPES = [
-        'int' => true,
-        'bool' => true,
-        'float' => true,
-        'string' => true,
-    ];
+    private NormalizerInterface $normalizer;
+
+    private DenormalizerInterface $denormalizer;
 
     protected ChainEncoder $encoder;
 
     protected ChainDecoder $decoder;
 
-    /**
-     * @var array<string, array<string, array<bool>>>
-     */
-    private array $denormalizerCache = [];
-
-    /**
-     * @var array<string, array<string, array<bool>>>
-     */
-    private array $normalizerCache = [];
-
     /**
      * @param array<NormalizerInterface|DenormalizerInterface> $normalizers
      * @param array<EncoderInterface|DecoderInterface>         $encoders
      */
     public function __construct(
-        private array $normalizers = [],
+        array $normalizers = [],
```

Comment:
Yeah. I dont like the current signature too. 

I want to have a signature like: 

```php
new Serializer($encoders, $normalizer, $denormalizer);
// or
`new Serializer($normalizer, $denormalizer, $encoders);`
```

Should I immediately go to that signature or should we change it in SF8?

---------

On your separate topic of making the `Chain*` classes an implementation detail.. That was an earlier version of the PR, but we decided against that to make the `Serializer` more simple. See https://github.com/symfony/symfony/pull/52900#discussion_r1417771466

Treating them like `$encoders` is a valid idea. However, you have 1-3 encoders MAX. But you may have 100+ normalizers/denormalizers. So I think it makes more sense to keep them separate and give the developer more control.

Reply:
> Should I immediately go to that signature or should we change it in SF8?

You **must** support the new signature immediately. Otherwise doing the change in 8.0 will be forbidden as it would be a BC break without migration path.

> On your separate topic of making the Chain* classes an implementation detail.. That was an earlier version of the PR, but we decided against that to make the Serializer more simple. See https://github.com/symfony/symfony/pull/52900#discussion_r1417771466

That comment was actually suggesting to _also_ change the way encoders are injected (we should avoid making 2 successive signature changes on the constructor, as that would be a pain to have 2 deprecated signatures to support in the BC layer)

> Treating them like $encoders is a valid idea. However, you have 1-3 encoders MAX. But you may have 100+ normalizers/denormalizers. So I think it makes more sense to keep them separate and give the developer more control.

but then, you make it more likely to have nested ChainNormalizer instance, making it necessary to be 100% sure about the reliability of `getSupportedTypes for instance (my suggested `return ['*' => false]` will be reliable, even though it might not be the most optimized one).